{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0",
   "metadata": {},
   "source": [
    "# Statistical Analysis of Differential Expression\n",
    "\n",
    "## Sample-Level Generalized Linear Mixed Models (GLMM) \n",
    "### see ../meta_analysis_GLMMs/glmm_fine_20250112.ipynb  and ../meta_analysis_GLMMs/glmm_coarse_20250703.ipynb \n",
    "\n",
    "To identify cluster-specific marker genes while accounting for spatial and technical sources of variation, we performed a Generalized Linear Mixed Model (GLMM) analysis separately for each biological sample. For each sample, raw gene counts were modeled using a Poisson error distribution with a canonical log link function. The model specification included the total number of UMIs (log-transformed) as an offset to normalize for sequencing depth. To capture spatial dependencies, we included random intercepts for the coarse k-nearest neighbor (k-NN) clusters (knn_coarse) and the field of view (fov), nested as (1 | fov / knn_coarse). This hierarchical structure accounts for the correlation of gene expression within spatial neighborhoods and imaging fields. Marginal effects for each cluster were estimated using the presto package, identifying genes with differential expression relative to the grand mean. Genes expressed in fewer than 3 cells per group were excluded prior to modeling.\n",
    "\n",
    "## DerSimonian-Laird Meta-Analysis\n",
    "\n",
    "### ../meta_analysis_cluster_markers/meta_analysis_of_fine_type_markers.ipynb \n",
    "\n",
    "To identify robust consensus markers across all biological replicates, sample-level summary statistics (log-transformed coefficients $\\beta$ and standard errors $\\sigma$) were combined using a random-effects meta-analysis. We employed the DerSimonian-Laird (DL) estimator to calculate the between-study variance ($\\tau^2$). For each feature within each cluster, we computed two estimates:Fixed Effects (FE) Model: Assumes a single true effect size shared by all samples, weighting estimates by the inverse of their within-sample variance ($w_{FE} = 1/\\sigma^2$).Random Effects (RE) Model: Incorporates both within-sample variance and between-sample heterogeneity ($\\tau^2$), weighting estimates as $w_{RE} = 1/(\\sigma^2 + \\tau^2)$.Heterogeneity was assessed using Cochranâ€™s $Q$ statistic. To avoid false positives driven by outlier samples, final feature prioritization was based on the Random Effects Z-score ($Z_{RE} = \\beta_{RE} / \\sigma_{RE}$), which penalizes genes with high inter-sample disagreement. P-values were adjusted for multiple hypothesis testing using the Benjamini-Hochberg (FDR) procedure. Features were considered significant if they exhibited an FDR < 0.05 in the Random Effects model.\n",
    "\n",
    "## Key References:\n",
    "\n",
    "- GLMM/Presto: https://rdrr.io/github/immunogenomics/presto/f/vignettes/getting-started.Rmd\n",
    "- Meta-Analysis: DerSimonian, R., & Laird, N. (1986). Meta-analysis in clinical trials. Controlled clinical trials, 7(3), 177-188."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "require(tidyverse)\n",
    "require(data.table)\n",
    "require(Matrix)\n",
    "require(singlecellmethods)\n",
    "require(lme4)\n",
    "require(presto)\n",
    "require(furrr)\n",
    "require(future)\n",
    "set.seed(1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "doGLMM = function(obj, filename) {\n",
    "    require(tidyverse)\n",
    "    require(data.table)\n",
    "    require(Matrix)\n",
    "    require(singlecellmethods)\n",
    "    require(lme4)\n",
    "    require(presto)\n",
    "    require(furrr)\n",
    "    require(future)\n",
    "    set.seed(1)\n",
    "  temp = GetAssayData(obj, slot = 'counts')\n",
    "  varyingGenes = rownames(temp[apply(temp, 1, function(x) {\n",
    "    length(unique(x)) > 3\n",
    "  }), ])\n",
    "  rm(temp)\n",
    "  obj = obj[varyingGenes,]\n",
    "  pb = presto::collapse_counts(\n",
    "    GetAssayData(obj, slot = 'counts'),\n",
    "    obj@meta.data,\n",
    "    c('orig.ident', 'fov', 'knn_renamed_cell_states'),\n",
    "    min_cells_per_group = 3\n",
    "  )\n",
    "  \n",
    "  pb$exprs_norm = pb$exprs_norm[rownames(pb$counts_mat), colnames(pb$counts_mat)]\n",
    "  \n",
    "  system.time({\n",
    "    presto_res = presto::presto.presto(\n",
    "      y ~ 1 + (1 |\n",
    "                 knn_renamed_cell_states) +  (1 | fov / knn_renamed_cell_states) + offset(logUMI),\n",
    "      pb$meta_data,\n",
    "      pb$counts_mat,\n",
    "      size_varname = \"logUMI\",\n",
    "      effects_cov = 'knn_renamed_cell_states',\n",
    "      ncore = 10,\n",
    "      min_sigma = .05,\n",
    "      family = \"poisson\",\n",
    "      nsim = 1000\n",
    "    )\n",
    "  })\n",
    "  \n",
    "  readr::write_rds(presto_res, filename)\n",
    "  \n",
    "  contrasts_mat = make_contrast.presto(presto_res,\n",
    "                                       var_contrast = \"knn_renamed_cell_states\")\n",
    "  \n",
    "  effects_marginal = contrasts.presto(presto_res,\n",
    "                                      contrasts_mat,\n",
    "                                      one_tailed = TRUE) %>%\n",
    "    dplyr::mutate(cluster = contrast) %>%\n",
    "    dplyr::mutate(\n",
    "      logFC = sign(beta) * log2(exp(abs(beta))),\n",
    "      # convert stats to log2 for interpretability\n",
    "      SD = log2(exp(sigma)),\n",
    "      zscore = logFC / SD\n",
    "    ) %>%\n",
    "    arrange(pvalue)\n",
    "  \n",
    "  effects_marginal$fdr = p.adjust(effects_marginal$pvalue, method = 'BH')\n",
    "  effects_marginal$corr_fdr = effects_marginal$fdr\n",
    "  effects_marginal$corr_fdr[effects_marginal$fdr == 0] = min(effects_marginal$fdr[effects_marginal$fdr != 0])\n",
    "  effects_marginal$`-log10_fdr` = (-1) * log10(effects_marginal$corr_fdr)\n",
    "  \n",
    "  # meanExp = rowMeans(GetAssayData(obj, 'data'))\n",
    "  # meanExp = data.frame(feature = names(meanExp), meanExp = meanExp)\n",
    "  # for (cluster in unique(effects_marginal$cluster)) {\n",
    "  #   temp = GetAssayData(obj, 'counts')[, rownames(obj@meta.data)[obj@meta.data$knn_renamed_cell_states == cluster]] %>% as.data.frame()\n",
    "  #   temp = temp %>%\n",
    "  #     rowwise() %>%\n",
    "  #     mutate(`N_zeros` = sum(c_across(everything()) == 0)) %>%\n",
    "  #     select(`N_zeros`) %>% as.data.frame()\n",
    "  #   rownames(temp) = rownames(GetAssayData(obj, 'counts'))\n",
    "  #   meanExp[, cluster] = temp$`N_zeros` / length(rownames(obj@meta.data)[obj@meta.data$knn_renamed_cell_states == cluster])\n",
    "  # }\n",
    "  return(effects_marginal)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "getwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "merfish_files = list.files(path = '../../../../Labeled MERFISH data/Seurat objects', pattern = \"annotated_.*.rds\", full.names = TRUE)\n",
    "length(merfish_files)\n",
    "merfish_files %>% writeLines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "names(merfish_files) = gsub(merfish_files, pattern = '.*annotated_|.rds', replacement = '')\n",
    "merfish_files"
   ]
  },
  {
   "cell_type": "raw",
   "id": "6",
   "metadata": {
    "tags": []
   },
   "source": [
    "sampleID = 'Jax001'\n",
    "obj = readr::read_rds(merfish_files[sampleID])\n",
    "obj"
   ]
  },
  {
   "cell_type": "raw",
   "id": "7",
   "metadata": {
    "tags": []
   },
   "source": [
    "lineages = unique(obj@meta.data$knn_coarse)\n",
    "lineages = lineages[!lineages %in% c('Mast', 'Plasma')]\n",
    "lineages"
   ]
  },
  {
   "cell_type": "raw",
   "id": "8",
   "metadata": {
    "tags": []
   },
   "source": [
    "lineage = 'Myeloid'\n",
    "obj_lineage = subset(obj, subset = knn_coarse == lineage)\n",
    "obj_lineage"
   ]
  },
  {
   "cell_type": "raw",
   "id": "9",
   "metadata": {
    "tags": []
   },
   "source": [
    "filename = paste0(c(sampleID, lineage, 'glmm.rds'), collapse = '_')\n",
    "system.time({\n",
    "    effects_marginal = doGLMM(obj = obj_lineage, filename = filename)\n",
    "})\n",
    "data.table::fwrite(effects_marginal, gsub(filename, pattern = 'glmm.rds', replacement = 'marginal_effects.csv'))"
   ]
  },
  {
   "cell_type": "raw",
   "id": "10",
   "metadata": {
    "tags": []
   },
   "source": [
    "head(effects_marginal)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "require(furrr)\n",
    "require(future)\n",
    "require(singlecellmethods)\n",
    "require(Seurat)\n",
    "set.seed(1)\n",
    "\n",
    "plan(multisession, workers = 10)\n",
    "future_map(names(merfish_files), function(sampleID){\n",
    "    message(sampleID)\n",
    "    obj = readr::read_rds(merfish_files[sampleID])\n",
    "    lineages = unique(obj@meta.data$knn_coarse)\n",
    "    lineages = lineages[!lineages %in% c('Mast', 'Plasma')]\n",
    "    lapply(lineages, function(lineage, sampleID){\n",
    "        message(lineages)\n",
    "        obj_lineage = subset(obj, subset = knn_coarse == lineage)\n",
    "        filename = paste0(c(sampleID, lineage, 'glmm.rds'), collapse = '_')\n",
    "        system.time({\n",
    "            effects_marginal = doGLMM(obj = obj_lineage, filename = filename)\n",
    "        })\n",
    "        data.table::fwrite(effects_marginal, gsub(filename, pattern = 'glmm.rds', replacement = 'marginal_effects.csv'))\n",
    "    }, sampleID = sampleID)\n",
    "}, .progress = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R Analysis Env (conda)",
   "language": "R",
   "name": "r-analysis-env"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
